File: Java/Tree/0199-binary-tree-right-side-view.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private void _helperRecirsiveDFS(TreeNode node, int level, List<Integer> result) {
          if (node == null) return;
          if (result.size() == level) {
              result.add(node.val);
          }
          _helperRecirsiveDFS(node.right, level + 1, result);
          _helperRecirsiveDFS(node.left, level + 1, result);
    }

    public List<Integer> rightSideView(TreeNode root) {
        if (root == null) return new ArrayList<>();

        if (root.left == null && root.right == null) {
            List<Integer> res = new ArrayList<>();
            res.add(root.val);
            return res;
        }
        
        List<Integer> result = new ArrayList<>();
        _helperRecirsiveDFS(root, 0, result);
        return result;
    }
}

File: Swift/Tree/0199-binary-tree-right-side-view.swift
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public var val: Int
 *     public var left: TreeNode?
 *     public var right: TreeNode?
 *     public init() { self.val = 0; self.left = nil; self.right = nil; }
 *     public init(_ val: Int) { self.val = val; self.left = nil; self.right = nil; }
 *     public init(_ val: Int, _ left: TreeNode?, _ right: TreeNode?) {
 *         self.val = val
 *         self.left = left
 *         self.right = right
 *     }
 * }
 */
class Solution {
    func rightSideView(_ root: TreeNode?) -> [Int] {
        if (root == nil) {
            return []
        }
        
        var rightSideView: [Int] = Array()
        
        var bfsQueue: [[TreeNode]] = Array()
        bfsQueue.append([root!])
        
        while (!bfsQueue.isEmpty) {
            var itemFromBfsQueue = bfsQueue.removeFirst()
            
            rightSideView.append(itemFromBfsQueue.last!.val)
            
            var newPackingarray: [TreeNode] = Array()
            
            for node in itemFromBfsQueue {
                if (node.left != nil) {
                    newPackingarray.append(node.left!)
                }
                
                if (node.right != nil) {
                    newPackingarray.append(node.right!)
                }
            }
            
            if (!newPackingarray.isEmpty) {
                bfsQueue.append(newPackingarray)
            }
        }
        return  rightSideView
    }
}